File: smpm-incompressible-navier-stokes-fourier-transverse/derv.f90
subroutine derv( nterm, x, d, d2, d3, ndim )
! This code contains the routine employeed to calculate the first, second, and
! third derivative matrix at Gauss-Lobatto-Legendre grid points calculated in
! the routines jacobl and jacobf.
!
! The theory behind this code can be found in Costa & Don "On the computation
! of high order derivatives".  Applied Numerical Mathematics. 33 (2000) 151 -
! 159
!
!   List of Variables:
!   nterm ->    Polynomial degree
!   x ->        Array with Gauss-Lobatto-Legendre Points
!   d ->        First derivative matrix
!   d2 ->       Second derivetive matrix
!   d3 ->       Third derivative matrix
!   ndim ->     Number of grid points in each direction

  use precision, only: dp

  implicit none

  ! Constants.
  integer, parameter                                    :: nn = 1000

  ! Subtroutine parameters.
  integer, intent(in)                                   :: nterm
  real(kind=dp), dimension(0:ndim), intent(in)          :: x
  real(kind=dp), dimension(0:ndim, 0:ndim), intent(out) :: d
  real(kind=dp), dimension(0:ndim, 0:ndim), intent(out) :: d2
  real(kind=dp), dimension(0:ndim, 0:ndim), intent(out) :: d3
  integer, intent(in)                                   :: ndim

  ! Local variables.
  real(kind=dp), dimension(0:nn)                        :: c
  real(kind=dp)                                         :: prod, sum

  integer                                               :: j, k, l     ! Loop indices.
  real(kind=dp)                                         :: xj, xk, xl  ! Temporary variables holding
                                                                       ! the jth, kth and lth,
                                                                       ! elements of x.

  ! Introduce methodology by Costa and Don.  see Eqn (4)
  ! 1) Must first compute coefficients C_i
  do k = 0, nterm
     prod = 1.0_dp
     xk   = x(k)

     do l = 0,nterm
        xl = x(l)

        if (l .ne. k) then
           prod = prod * (xk - xl)
        else
           prod = prod
        endif
     enddo

     c(k) = prod
  enddo

  ! First Derivative:
  ! Calculate off diagonal elements of D^1.  see Eqn (6)
  do k = 0, nterm
     xk = x(k)

     do j = 0, nterm
        xj = x(j)

        if (k .ne. j) then
           d(k, j) = c(k) / (c(j) * (xk - xj))
        else
           d(k, j) = 0.0_dp
        endif
     enddo
  enddo

   ! Calculate diagonal elements now.  Diagonal element is negative row sum of
   ! off diagonal elements (See Costa & Don) Eqn(9)
   do k = 0, nterm
      sum = 0.0_dp

      do j = 0, nterm
         if (k .ne. j) then
            sum = sum + d(k, j)
         else
            sum = sum
         endif
      enddo

      d(k, k) = -sum
   enddo

   ! Second derivative:
   ! Off-diagonal elements Eqn(13)
   do k = 0, nterm
      xk = x(k)

      do j = 0, nterm
         xj = x(j)

         if (k .ne. j) then
            d2(k, j) = 2.0_dp * (d(k, k) * d(k, j) - d(k, j) / (xk - xj))
         else
            d2(k, j) = 0.0_dp
         endif
      enddo
   enddo

   ! Diagonal elements Eqn(9) again
   do k = 0, nterm
      sum = 0.0_dp

      do j = 0, nterm

         if (k .ne. j) then
            sum = sum + d2(k, j)
         else
            sum = sum
         endif
      enddo

      d2(k, k) = -sum
   enddo

   ! Third derivative:
   ! Off-diagonal elements
   do k = 0, nterm
      xk = x(k)

      do j = 0, nterm
         xj = x(j)

         if (k .ne. j) then
            d3(k, j) = 3.0_dp * (d2(k, k) * d(k, j) - d2(k, j) / (xk - xj))
         else
            d3(k, j) = 0.0_dp
         endif
      enddo
   enddo

    ! Diagonal elements
    do k = 0, nterm
       sum = 0.

       do j = 0, nterm
          if (k .ne. j) then
             sum = sum + d3(k, j)
          else
             sum = sum
          endif
       enddo

       d3(k, k) = -sum
    enddo

return

end subroutine derv

File: smpm-incompressible-navier-stokes-fourier-transverse/smpm_matlab_utilities/derv.m
function [d, d2, d3] = derv( nterm, x, ndim )
% [d, d2, d3] = derv( nterm, x, ndim )
%
% Calculate the first, second, and third derivative matrix at
% Gauss-Lobatto-Legendre grid points calculated in the Fortran routines
% jacobl() and jacobf().
%
% The theory behind this code can be found in Costa & Don "On the computation
% of high order derivatives".  Applied Numerical Mathematics. 33 (2000) 151 -
% 159.
%
% Takes 3 arguments:
%
%   nterm - Polynomial degree.
%   x     - Gauss-Lobatto-Legendre Points.
%   ndim  - Number of grid points in each direction.
%
% Returns 3 values:
%
%   d  - Vector, length (n+1)*(n+1), of first derivatives in column-wise order.
%   d2 - Vector, length (n+1)*(n+1), of second derivatives in column-wise order.
%   d3 - Vector, length (n+1)*(n+1), of third derivatives in column-wise order.

% Adapted from Pete Diamessis' derv.f and Sumedh Joshi's derv.f90.
%
% Gustavo Rivera 2016

% Allocate derivative matrices.
d  = zeros( ndim+1, ndim+1 );
d2 = zeros( ndim+1, ndim+1 );
d3 = zeros( ndim+1, ndim+1 );

% 1) First compute coefficients C_i.  See Eqn (4) in Costa and Don.
for k = 1: nterm+1
    prod = 1.;
    xk   = x(k);

    for l = 1:nterm+1
        xl = x(l);

        if l ~= k
            prod = prod * (xk - xl);
        else
            prod = prod;
        end
    end

    c(k) = prod;
end

%% Compute the first derivative.

% Calculate off-diagonal elements of D^1.  See Eqn (6).
for k = 1: nterm+1
    xk = x(k);

    for j = 1:nterm+1
        xj = x(j);

        if k ~= j
            d(k, j) = c(k) / (c(j) * (xk - xj));
        else
            d(k, j) = 0.;
        end
    end
end

% Calculate diagonal elements now.  Diagonal element is negative row sum of
% off diagonal elements.  See Costa & Don Eqn(9).
for k = 1: nterm+1
    sum = 0.;

    for j = 1: nterm+1
        if k ~= j
            sum = sum + d(k, j);
        else
            sum = sum;
        end
    end

    d(k, k) = -sum;
end

%% Compute the second derivative.

% Off-diagonal elements.  See Eqn(13).
for k = 1: nterm+1
    xk = x(k);

    for j = 1: nterm+1
        xj = x(j);

        if k ~= j
            d2(k, j) = 2. * (d(k, k) * d(k, j) - d(k, j) / (xk - xj));
        else
            d2(k, j) = 0.;
        end
    end
end

% Diagonal elements.  See Eqn(9).
for k = 1: nterm+1
    sum = 0.;

    for j = 1:nterm+1

        if k ~= j
            sum = sum + d2(k, j);
        else
            sum = sum;
        end
    end

    d2(k, k) = -sum;
end

%% Compute the third derivative.

% Off-diagonal elements.
for k = 1: nterm+1
    xk = x(k);

    for j = 1: nterm+1
        xj = x(j);

        if k ~= j
            d3(k, j) = 3. * (d2(k, k) * d(k, j) - d2(k, j) / (xk - xj));
        else
            d3(k, j) = 0.;
        end
    end
end

% Diagonal elements.
for k = 1: nterm+1
    sum = 0.;

    for j = 1: nterm+1
        if k ~= j
            sum = sum + d3(k, j);
        else
            sum = sum;
        end
    end

    d3(k, k) = -sum;
end

% Reverse the ordering of the matrices and flatten them into vectors.
d  = flipud(  d(:) );
d2 = flipud( d2(:) );
d3 = flipud( d3(:) );

end
